backend/corpora/common/entities/collection.py

```diff
@@ -245,23 +244,49 @@ def publish(self, data_submission_policy_version):
         # Timestamp for published_at and revised_at
         now = datetime.utcnow()
         # Create a public collection with the same uuid and same fields
-        is_existing_collection = False
         if self.revision_of:
             # This is a revision of a published Collection
             public_collection = Collection.get_collection(self.session, collection_uuid=self.revision_of)
+            has_dataset_changes = False
+            for dataset in self.datasets:
+                revised_dataset = Dataset(dataset)
+                original = (
+                    Dataset.get(self.session, revised_dataset.original_id) if revised_dataset.original_id else None
+                )
+                if original and public_collection.check_has_dataset(original):
+                    dataset_is_changed = original.publish_revision(revised_dataset, now)
+                    if dataset_is_changed:
+                        has_dataset_changes = True
+                else:
+                    # The dataset is new
+                    revised_dataset.publish_new(now)
+                    has_dataset_changes = True
+            self.session.flush()
+            # calling expire because the above code updates and deletes rows related to this object. If we don't expire
+            # the object, SqlAlchmey will try to modify rows that don't exists and thrown an error.
+            self.session.expire(self.db_object)
             revision = self.to_dict(
-                remove_attr=("updated_at", "created_at", "visibility", "id", "revision_of", "published_at"),
+                remove_attr=(
+                    "updated_at",
+                    "created_at",
+                    "visibility",
+                    "id",
+                    "revision_of",
+                    "published_at",
+                    "revised_at",
+                ),
                 remove_relationships=True,
             )
             revision["data_submission_policy_version"] = data_submission_policy_version
+            revision["revised_at"] = now if has_dataset_changes else self.revised_at
             public_collection.update(
-                commit=False,
-                **revision,
-            )
+                **revision, commit=False
+            )  # This function call deletes old links (keep_links param defaults to False)
             for link in self.links:
-                CollectionLink(link).update(collection_id=self.revision_of, commit=False)
-            is_existing_collection = True
-
+                link.collection_id = self.revision_of
+            self.session.flush()
```

Comment:
@Bento007 I just added this line. I am pretty sure this is necessary to make 100% sure that the links are moved to the published object rather than themselves deleted. I think we had an inverse of this situation with the original bug which meant that the code usually produced the bug but occasionally worked (as per Jason's experience); without this line, I think the code would usually work but could possibly produce the bug.

I think the combination of reassigning the links and then the subsequent `delete` call does not have deterministic behavior with SQLAlchemy.

Basically, the flush is required to make 100% sure that the db transaction first sees that the revision links now belong to the published collection _before_ the revision object is deleted (in which case the links would themselves be deleted).